gateway/uplink.go

```diff
@@ -0,0 +1,148 @@
+package gateway
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/robertkrimen/otto"
+	"go.thethings.network/lorawan-stack/v3/pkg/crypto"
+	"go.thethings.network/lorawan-stack/v3/pkg/types"
+)
+
+// Structure of phyPayload:
+// | MHDR | DEV ADDR|  FCTL |   FCnt  | FPort   |  FOpts     |  FRM Payload | MIC |
+// | 1 B  |   4 B    | 1 B   |  2 B   |   1 B   | variable    |  variable   | 4B  |
+func (g *Gateway) parseDataUplink(phyPayload []byte) (string, map[string]interface{}, error) {
```

Comment:
are your helpers easy to set up tests for at this point?

Reply:
I will write tests after the split of device/gateway to avoid having to change the structure of the tests